pytorch_lightning/utilities/distributed.py

```diff
@@ -62,13 +65,41 @@ def gather_all_tensors(result: torch.Tensor, group: Optional[Any] = None) -> Lis
     result = result.contiguous()
 
     world_size = torch.distributed.get_world_size(group)
+    torch.distributed.barrier(group=group)
 
-    gathered_result = [torch.zeros_like(result) for _ in range(world_size)]
+    # if the tensor is scalar, things are easy
+    if result.ndim == 0:
+        return _simple_gather_all_tensors(result, group, world_size)
+
+    # 1. Gather sizes of all tensors
+    local_size = torch.tensor(result.shape, device=result.device)
+    local_sizes = [torch.zeros_like(local_size) for _ in range(world_size)]
+    torch.distributed.all_gather(local_sizes, local_size, group=group)
+    max_size = torch.stack(local_sizes).max(dim=0).values
+    all_sizes_equal = all(all(ls == max_size) for ls in local_sizes)
```

Comment:
n00b question, is this only used by metrics? Is that possible that user use `gather_all_tensors` in each steps for some other purpose? If yes, this will add one more collective communication for every all_gather, do we want to make this optional?

Reply:
Have a similar question as well, should size-checking and padding become part of the default in `gather_all_tensors`? 

Maybe we can add as an option which could be turned on/off instead, so that it avoid some of the complexities and overheads if not necessary?